### PR TITLE
core: add ExtensionDirectories= setting

### PR DESCRIPTION
**Pull request by bluca**
Add a new setting that follows the same principle and implementation
as ExtensionImages, but using directories as sources.
It will be used to implement support for extending portable images
with directories, since portable services can already use a directory
as root.

### CODE REVIEW

File: man/systemd.exec.xml [hunk 1]
```diff
@@ -459,6 +459,35 @@
...
+        up. This option expects a whitespace separated list of source directories.</para>
+
+        <para>A read-only OverlayFS will be set up on top of <filename>/usr/</filename> and
+        <filename>/opt/</filename> hierarchies. The order in which the directories are listed will determine
+        the order in which the overlay is laid down: directories specified first to last will result in overlayfs
+        layers bottom to top.</para>
+
+        <para>Each directory may be prefixed with <literal>-</literal>, in which case it will be ignored
+        when its source path does not exist. If the source path contains a <literal>:</literal>, it needs to be
+        escaped as <literal>\:</literal>. Any mounts created with this option are specific to the unit, and are
```
Comment (yuwata) [inline man/systemd.exec.xml]: It looks the colon is automatically escaped.

File: src/core/namespace.c [hunk 2]
```diff
@@ -482,6 +484,62 @@ static int append_extension_images(
...
+                        if (!prefixed_hierarchy)
+                                return -ENOMEM;
+
+                        escaped = shell_escape(prefixed_hierarchy, ",:");
+                        if (!escaped)
+                                return -ENOMEM;
+
+                        /* Note that lowerdir= parameters are in 'reverse' order, so the
+                         * top-most directory in the overlay comes first in the list. */
+                        lowerdir = strjoin(escaped, ":", overlays[j]);
```
Comment (yuwata) [inline src/core/namespace.c:526]: Does overlayfs gracefully handle ENOENT?

File: src/core/namespace.c [hunk 3]
```diff
@@ -482,6 +484,62 @@ static int append_extension_images(
...
+                        return -ENOMEM;
+
+                /* Look for any prefixes */
+                if (startswith(e, "-")) {
+                        e++;
+                        ignore_enoent = true;
+                }
+                /* Ignore this for now */
+                if (startswith(e, "+"))
+                        e++;
```
Comment (yuwata) [inline src/core/namespace.c:507]: Is this necessary?

Comment (bluca) [reply src/core/namespace.c:507]: Yes, the parser allows it - might add support for it later

File: src/core/namespace.c (hunk 2 above)
Comment (bluca) [reply src/core/namespace.c:526]: It does not, but this is an existing issue with ExtensionImages too and will need a common solution, which will be a tad convoluted

File: test/units/testsuite-50.sh [hunk 4]
```diff
@@ -321,6 +321,34 @@ EOF
...
+systemd-dissect --mount /usr/share/app0.raw "${image_dir}/app0"
+systemd-dissect --mount /usr/share/app1.raw "${image_dir}/app1"
+systemd-run -P --property ExtensionDirectories="${image_dir}/app0" --property RootImage="${image}.raw" cat /opt/script0.sh | grep -q -F "extension-release.app0"
+systemd-run -P --property ExtensionDirectories="${image_dir}/app0" --property RootImage="${image}.raw" cat /usr/lib/systemd/system/some_file | grep -q -F "MARKER=1"
+systemd-run -P --property ExtensionDirectories="${image_dir}/app0 ${image_dir}/app1" --property RootImage="${image}.raw" cat /opt/script0.sh | grep -q -F "extension-release.app0"
+systemd-run -P --property ExtensionDirectories="${image_dir}/app0 ${image_dir}/app1" --property RootImage="${image}.raw" cat /usr/lib/systemd/system/some_file | grep -q -F "MARKER=1"
+systemd-run -P --property ExtensionDirectories="${image_dir}/app0 ${image_dir}/app1" --property RootImage="${image}.raw" cat /opt/script1.sh | grep -q -F "extension-release.app2"
+systemd-run -P --property ExtensionDirectories="${image_dir}/app0 ${image_dir}/app1" --property RootImage="${image}.raw" cat /usr/lib/systemd/system/other_file | grep -q -F "MARKER=1"
```
Comment (yuwata) [inline test/units/testsuite-50.sh:335]: Maybe better to add tests for empty directory and nonexistent directory cases.

File: man/systemd.exec.xml (hunk 1 above)
Comment (bluca) [reply man/systemd.exec.xml]: yes, copied the paragraph from above - removed

File: src/core/dbus-execute.c [hunk 5]
```diff
@@ -3291,6 +3293,8 @@ int bus_exec_context_set_transient_property(
                                 dirs = &c->exec_paths;
                         else if (streq(name, "NoExecPaths"))
                                 dirs = &c->no_exec_paths;
+                        else if (streq(name, "ExtensionDirectories"))
+                                dirs = &c->extension_directories;
```
Comment (yuwata) [inline src/core/dbus-execute.c:3297]: Hmm, there is no validity check...

Comment (bluca) [reply src/core/dbus-execute.c:3297]: Same as other directory-based parameters, there's a minimal parsing and validity checks done by the parser (absolute path, etc), but beyond that it's checked at use-time, not at configure-time

File: man/systemd.exec.xml [hunk 6]
```diff
@@ -459,6 +459,34 @@
...
+        <term><varname>ExtensionDirectories=</varname></term>
+
+        <listitem><para>This setting is similar to <varname>BindReadOnlyPaths=</varname> in that it mounts a file
+        system hierarchy from a directory, but instead of providing a destination path, an overlay will be set
+        up. This option expects a whitespace separated list of source directories.</para>
+
+        <para>A read-only OverlayFS will be set up on top of <filename>/usr/</filename> and
+        <filename>/opt/</filename> hierarchies. The order in which the directories are listed will determine
+        the order in which the overlay is laid down: directories specified first to last will result in overlayfs
+        layers bottom to top.</para>
```
Comment (yuwata) [inline man/systemd.exec.xml:473]: Does this paragraph implies that /usr and /opt must exist under the directory unless prefixed with '-'?

File: src/core/namespace.c (hunk 3 above)
Comment (yuwata) [reply src/core/namespace.c:507]: Indeed.

File: man/systemd.exec.xml (hunk 6 above)
Comment (bluca) [reply man/systemd.exec.xml:473]: No it applies to the setting - reworded to clarify

File: test/units/testsuite-50.sh (hunk 4 above)
Comment (bluca) [reply test/units/testsuite-50.sh:335]: added both

File: src/core/dbus-execute.c (hunk 5 above)
Comment (yuwata) [reply src/core/dbus-execute.c:3297]: Oops. Each string is checked in the above. Sorry for the noise. Please forget.

**Review by yuwata — APPROVED** [review]

**Pull request merged**